pymc_experimental/utils/model_fgraph.py

```diff
@@ -0,0 +1,174 @@
+
+    def infer_shape(self, fgraph, node, inputs_shape):
+        return inputs_shape[0]
+
+    def do_constant_folding(self, fgraph, node):
+        return False
+
+    def perform(self, *args, **kwargs):
+        raise RuntimeError("ValuedRVs should never be evaluated!")
+
+
+class FreeRV(ModelVar):
+    pass
+
+
+class ObservedRV(ModelVar):
+    pass
+
+
+class Potential(ModelVar):
+    pass
+
+
+class Deterministic(ModelVar):
+    pass
+
+
+free_rv = FreeRV()
+observed_rv = ObservedRV()
+potential = Potential()
+deterministic = Deterministic()
+
+
+def toposort_replace(fgraph: FunctionGraph, replacements) -> None:
+    """Replace multiple variables in topological order."""
+    toposort = fgraph.toposort()
+    sorted_replacements = sorted(replacements, key=lambda pair: toposort.index(pair[0].owner))
+    fgraph.replace_all(tuple(sorted_replacements), import_missing=True)
+
+
+def fgraph_from_model(model: Model) -> FunctionGraph:
+
+    # Collect PyTensor variables
+    rvs_to_values = model.rvs_to_values
+    rvs = list(rvs_to_values.keys())
+    values = list(rvs_to_values.values())
+    free_rvs = model.free_RVs
+    deterministics = model.deterministics
+    potentials = model.potentials
+
+    # Collect PyMC meta-info
+    vars_to_dims = model.named_vars_to_dims
+    coords = model.coords
+
+    # TODO: Do something with these
+    dim_lengths = model.dim_lengths
+    rvs_to_transforms = model.rvs_to_transforms
+
+    # Not supported yet
+    if any(v is not None for v in model.rvs_to_total_sizes.values()):
+        raise NotImplementedError("Cannot convert models with total_sizes")
+    if any(v is not None for v in model.rvs_to_initial_values.values()):
+        raise NotImplementedError("Cannot convert models with non-default initial_values")
+
+    # We start the `dict` with mappings from the value variables to themselves,
+    # to prevent them from being cloned.
+    memo = {v: v for v in values}
+
+    fgraph = FunctionGraph(
+        outputs=rvs + potentials + deterministics,
+        clone=True,
+        memo=memo,
+        copy_orphans=False,
+        copy_inputs=False,
+    )
+    fgraph.coords = coords
+
+    # Introduce dummy `ModelVar` Ops
+    free_rvs_to_values = {memo[k]: v for k, v in rvs_to_values.items() if k in free_rvs}
+    observed_rvs_to_values = {memo[k]: v for k, v in rvs_to_values.items() if k not in free_rvs}
+    potentials = [memo[k] for k in potentials]
+    deterministics = [memo[k] for k in deterministics]
+
+    vars = fgraph.outputs
+    new_vars = []
+    for var in vars:
+        dims = vars_to_dims.get(var.name, None)
+        if var in free_rvs_to_values:
+            new_var = free_rv(var, free_rvs_to_values[var], dims)
+        elif var in observed_rvs_to_values:
+            new_var = observed_rv(var, observed_rvs_to_values[var], dims)
+        elif var in potentials:
+            new_var = potential(var, dims)
+        elif var in deterministics:
+            new_var = deterministic(var, dims)
```

Comment:
We might want to pull Deterministics out of the graph in `Fgraph` (so they don't bother us during rewrites) and back into the graph in `Model` (so they print nicely in graphviz).